### PR TITLE
SNOW-3329144: Add Cortex Code CLI agent for autonomous E2E PR review

### PR DESCRIPTION
**Pull request by sfc-gh-olorek**
## SNOW-3329144

Adds an autonomous Cortex AI agent that performs end-to-end verification of PRs using Cortex Code CLI (powered by claude-opus-4-6).

When a PR changes CLI behavior, the agent autonomously investigates the codebase, executes CLI commands, runs SQL queries in an ephemeral playground database, verifies side effects, identifies breaking changes, and posts a structured review comment. For non-behavioral PRs (CI, docs, tests-only), it reports SKIP.

## How it works

Cortex Code CLI is the agent — it natively handles file reading, command execution, SQL queries, grep, and the agentic loop. The script (~300 lines) is thin orchestration: setup playground, run cortex, post comment, cleanup.

**Pipeline:**
1. Verify `snow` CLI build from PR branch
2. Install and verify Cortex Code CLI
3. Create ephemeral playground database (`CORTEX_REVIEW_PR<N>_<RUN_ID>`)
4. Configure Snowflake connection for Cortex
5. Run Cortex Code CLI agent with full autonomy (`--bypass --plan --auto-accept-plans`)
6. Delete previous bot comment, post new one with structured report
7. Drop playground database

**Investigation methodology** (via system prompt):
1. Understand scope — read changed files, grep for callers/imports
2. Test happy path — run changed CLI commands
3. Test edge cases — empty inputs, special chars, missing objects
4. Identify breaking changes — renamed commands, changed defaults, altered output formats, different exit codes
5. Check interactions — grep shared utilities, test related commands
6. Verify side effects — query Snowflake to confirm changes took effect
7. Report potential risks — flag things not testable in the playground

**Report sections:** Summary, E2E Test Results, Breaking Changes, Side-Effect Verification, Potential Risks, Verdict (PASS/FAIL/SKIP)

## Files
- `.github/workflows/cortex_review.yaml` — GitHub Actions workflow
- `.github/scripts/cortex_e2e_review.py` — Orchestration script

## Safety
- **Same-repo only** — fork PRs cannot trigger (no secrets exposure)
- **Ephemeral playground** — isolated database, dropped after run
- **Ephemeral runner** — CI runner destroyed after job
- **60-minute timeout** — workflow backstop
- **Comment deduplication** — deletes previous bot comment on re-push
- **Cleanup in finally** — playground always dropped, even on failure

## Prerequisites
```sql
ALTER ACCOUNT SET CORTEX_ENABLED_CROSS_REGION = 'ANY_REGION';
ALTER ACCOUNT SET CORTEX_MODELS_ALLOWLIST = 'All';
GRANT DATABASE ROLE SNOWFLAKE.CORTEX_USER TO ROLE PUBLIC;
```

### CODE REVIEW

File: .github/workflows/cortex_review.yaml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  pull-requests: write
+
+jobs:
+  cortex-review:
+    if: |
+      github.event.pull_request.head.repo.full_name == github.repository &&
+      github.event.pull_request.draft == false &&
+      contains(github.event.pull_request.labels.*.name, 'enable-ai-e2e-review') &&
+      (github.event.action != 'labeled' || github.event.label.name == 'enable-ai-e2e-review')
+    runs-on: ubuntu-latest
```
Comment (sfc-gh-kolszewski) [inline .github/workflows/cortex_review.yaml:31]: I think we wanted this functionality to be accessible only to maintainers for now so we should include some check on the PR author here.

Comment (sfc-gh-olorek) [reply .github/workflows/cortex_review.yaml:31]: Good call. Added an author gate to the job `if:` — it now requires `github.event.pull_request.author_association` to be `OWNER`, `MEMBER`, or `COLLABORATOR` (via `contains(fromJSON('["OWNER", "MEMBER", "COLLABORATOR"]'), ...)`), so only org members/collaborators can trigger the review for now — on top of the existing same-repo and opt-in label checks. Done in 5dfb0def.

Comment (sfc-gh-kolszewski) [reply .github/workflows/cortex_review.yaml:31]: Can we also have a check for who applied the label? Right now any CONTRIBUTOR (triage access) can add enable-ai-e2e-review to a trusted member's PR

Comment (sfc-gh-olorek) [reply .github/workflows/cortex_review.yaml:31]: Good catch — `author_association` only vouches for the PR author, not whoever applied the label.

Added an `authorize` gate job (the review job now `needs:` it) that checks the **triggering actor's** repo permission via the collaborator-permission API. On a `labeled` event `github.actor` is the user who applied the label, so this is exactly the "who applied the label" check. That endpoint reports triage access as `read`, so requiring `write`/`admin` excludes it — a triage-access contributor who adds the label fails the gate and the secret-bearing review job never starts. `set -e` makes any API error fail closed.

Done in 53e0deb8.

**Review by sfc-gh-kolszewski — APPROVED** [review]
Tested the recent changes with Maintainer authorization gate and changed credentials on https://github.com/snowflakedb/snowflake-cli/pull/3106
